src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,163 @@
+/* Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic {
+  namespace util {
+
+    template<class T> 
+    class SlidingWindow {
+      public:
+        SlidingWindow(UInt maxCapacity);
+        SlidingWindow(UInt maxCapacity, std::vector<T> initialData);
```

Comment:
Noting example how to write iterator, I have some trouble w it 
http://en.cppreference.com/w/cpp/iterator/iterator 
This `Iterator` should be our overloaded class?